File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.java
import java.util.*;

class Solution {
    public int solution(int m, int n, String[] board) {
        char[][] grid = new char[m][n];
        
        // board 입력을 char 배열로 변환
        for (int i = 0; i < m; i++) {
            grid[i] = board[i].toCharArray();
        }
        
        int totalRemoved = 0;
        
        while (true) {
            // 제거할 블록 찾기
            boolean[][] marked = new boolean[m][n];
            boolean found = false;
            
            for (int i = 0; i < m - 1; i++) {
                for (int j = 0; j < n - 1; j++) {
                    char c = grid[i][j];
                    if (c != ' ' && c == grid[i][j + 1] && c == grid[i + 1][j] && c == grid[i + 1][j + 1]) {
                        marked[i][j] = true;
                        marked[i][j + 1] = true;
                        marked[i + 1][j] = true;
                        marked[i + 1][j + 1] = true;
                        found = true;
                    }
                }
            }
            
            // 찾은 블록이 없다면 정지
            if (!found) break;
            
            // 블록 카운트 및 제거
            int removed = 0;
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (marked[i][j]) {
                        grid[i][j] = ' ';
                        removed++;
                    }
                }
            }
            totalRemoved += removed;
            
            // 블록 떨어뜨리기
            for (int j = 0; j < n; j++) {
                for (int i = m - 1; i >= 0; i--) {
                    if (grid[i][j] == ' ') {
                        // 다음 떨어질 블록 찾기
                        int k = i - 1;
                        while (k >= 0 && grid[k][j] == ' ') k--;
                        if (k >= 0) {
                            grid[i][j] = grid[k][j];
                            grid[k][j] = ' ';
                        }
                    }
                }
            }
        }
        
        return totalRemoved;
    }
}

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.py
def solution(m, n, board):
    answer = 0
    b = [list(x) for x in board]
    tmp = set()
    while True:
        for i in range(m-1):
            for j in range(n-1):
                t= b[i][j]
                if t==[]:
                    continue
                else:
                    if b[i][j+1]==t and b[i+1][j]==t and b[i+1][j+1]==t:
                        tmp.add((i, j))
                        tmp.add((i, j+1))
                        tmp.add((i+1, j))
                        tmp.add((i+1, j+1))
        if tmp:
            answer+=len(tmp)
            for i , j in tmp:
                b[i][j] = []
            tmp = set()
        else:
            break
            
        while True:
            moved=0
            for i in range(m-1):
                for j in range(n):
                    if b[i][j] and b[i+1][j] == []:
                        b[i][j],b[i+1][j] = b[i+1][j],b[i][j]
                        moved = 1
            if moved==0:
                break
                
    
    return answer